### PR TITLE
[#163][#191][#300]   Upgrade JDK, SDK, Gradle & Kotlin

### PR DESCRIPTION
**Pull request by doannimble**
https://github.com/nimblehq/android-templates/issues/163
https://github.com/nimblehq/android-templates/issues/191
https://github.com/nimblehq/android-templates/issues/300

## What happened 👀

- Java 1.8 is old, we should upgrade it to 11 in order to use latest Gradle or Kotlin version
- Upgrade Compile and Target SDK version from 30 to 33
- Update Gradle to version 7+

## Insight 📝

- Upgrade JDK 11 for CI Pull Request Review and Run Detekt/Unit Test
- Upgrade Gradle to 7.3.0
- Upgrade min Android from 23 -> 24
- Upgrade Java from 8 to 11

## Proof Of Work 📹

N/A

### CODE REVIEW

**Comment by sleepylee** [comment]
a quick reminder on these places:

https://github.com/nimblehq/android-templates/blob/feature/191-upgrade-jdk-11/template/build.gradle.kts#L34

https://github.com/nimblehq/android-templates/blob/feature/191-upgrade-jdk-11/template/app/build.gradle.kts#L84

https://github.com/nimblehq/android-templates/blob/feature/191-upgrade-jdk-11/template/app/build.gradle.kts#L88

....

**Comment by Tuubz** [comment]
@doannimble Any reason we don't include these changes with the [other pull request](https://github.com/nimblehq/android-templates/pull/322/files#diff-8d83febc7bad9355f66dcdf21f1a25e2b5ee979000e8b11b4446d880542f969bR83-R84)? 🙏

**Comment by doannimble** [comment]
> @doannimble Any reason we don't include these changes with the [other pull request](https://github.com/nimblehq/android-templates/pull/322/files#diff-8d83febc7bad9355f66dcdf21f1a25e2b5ee979000e8b11b4446d880542f969bR83-R84)? 🙏
> 
> One more question, if this is "Part 1", what can we expect from "Part 2"?

@Tuubz In this PR, I only need to update the JDK from CI build. Others related our code base will be implement in [`Part 2`](https://github.com/nimblehq/android-templates/pull/322)

**Comment by Tuubz** [comment]
@doannimble Thanks for the clarification 🙏 

Do you agree it would be better to combine both pull requests into one?

**Comment by doannimble** [comment]
@Tuubz Possible, we can merge the Part 2 into the Part 1, Let me do it 👍

**Comment by Tuubz** [comment]
@doannimble I'm afraid we can't merge the pull request like this (because code coverage has dropped to `0.66%`)

If someone wants to generate a project afterward, they will be given an incorrect code coverage

I'll create a ticket for Kover soon, so we can review this pull request again after (next sprint)  🚀

File: .github/workflows/verify_newproject_script.yml
Comment (Tuubz) [inline .github/workflows/verify_newproject_script.yml:18]: Wrong indentation, CI is failing because of this 🙏

File: .github/workflows/run_detekt_and_unit_tests.yml
Comment (Tuubz) [inline .github/workflows/run_detekt_and_unit_tests.yml:14]: Just wondering, is it required to specify `distribution: 'temurin'`?

File: template/buildSrc/src/main/java/Versions.kt
Comment (Tuubz) [inline template/buildSrc/src/main/java/Versions.kt:16]: Because the library is named `androidx.test:core:x.x.x`
```suggestion
    const val ANDROIDX_TEST_CORE_VERSION = "1.4.0"
```

File: template/build.gradle.kts
Comment (Tuubz) [inline template/build.gradle.kts:41]: Could it be this is redundant? Since we already define `tasks.withType<io.gitlab.arturbosch.detekt.Detekt>` on line 64

File: template/app/build.gradle.kts
Comment (Tuubz) [inline template/app/build.gradle.kts:136]: Any reason we use `runtimeOnly` instead of `implementation`? 💭

File: template/build.gradle.kts
Comment (Tuubz) [inline template/build.gradle.kts:33]: Would you mind giving some explanation on why this is required? 🙇

File: template/build.gradle.kts
Comment (Tuubz) [inline template/build.gradle.kts:78]: Any idea why we need this? 🙏

**Review by Tuubz — COMMENTED** [review]
@doannimble 🚀

**Comment by Tuubz** [comment]
@doannimble Oh, one more thing. Seems like something went wrong with rebasing, because `template/buildSrc/src/main/java/plugins/jacoco-report.gradle.kts` has been added again 🙏

File: .github/workflows/run_detekt_and_unit_tests.yml
Comment (doannimble) [reply .github/workflows/run_detekt_and_unit_tests.yml:14]: @Tuubz Yes it is https://github.com/actions/setup-java/issues/294

File: template/build.gradle.kts
Comment (doannimble) [reply template/build.gradle.kts:41]: @Tuubz Good catch, this is the redundant one 🙏

File: template/app/build.gradle.kts
Comment (doannimble) [reply template/app/build.gradle.kts:136]: @Tuubz We can use `runtimeOnly` to reduce the compile time, as this ktx can be use at Runtime
But it does not necessary in our template, I will use the implementation for consistency.

File: template/build.gradle.kts
Comment (doannimble) [reply template/build.gradle.kts:33]: @Tuubz It is used for Gradle cache change, my prasice when trying to resolve Jacoco coverage report. Now we dont need it anymore 🙏

File: template/build.gradle.kts
Comment (doannimble) [reply template/build.gradle.kts:78]: @Tuubz They're redundant, we only need `tasks.withType<io.gitlab.arturbosch.detekt.Detekt>` 🙏

**Review by Tuubz — COMMENTED** [review]
@doannimble Last comments 🙇

File: template/build.gradle.kts
Comment (Tuubz) [inline template/build.gradle.kts:58]: `Archive Detekt reports on template` in `run_detekt_and_unit_tests.yml` expects path `template/build/reports/detekt/`

```suggestion
            outputLocation.set(file("build/reports/detekt/detekt.xml"))
```

Shall we re-factor it like below? Because the default is `template/build/reports/detekt/` anyways
 
```
tasks.withType<io.gitlab.arturbosch.detekt.Detekt>().configureEach {
    jvmTarget = JavaVersion.VERSION_11.toString()
}
```

**Review by Tuubz — APPROVED** [review]
@doannimble Rest LGTM 🚀 

Could you make another pull request to apply the same changes to `sample-compose` & `sample-xml`? I have unlinked the issues from this pull request for now, otherwise they will get closed when merging this one 🙏

**Review by AVI5HEK — APPROVED** [review]

File: template/build.gradle.kts
Comment (Tuubz) [inline template/build.gradle.kts:61]: @doannimble 🤪 

```suggestion
            outputLocation.set(file("build/reports/detekt/detekt.html"))
```

**Review by Tuubz — APPROVED** [review]
LGTM 🔥 

If I'm not mistaken, the issues that have been integrated into this pull request are:

- #163 
- #191 
- #300 

Could you update the title, description, and commit history accordingly? 🙏

File: template/data/src/test/java/co/nimblehq/template/data/extensions/ResponseMappingTest.kt
Comment (chornerman) [inline template/data/src/test/java/co/nimblehq/template/data/extensions/ResponseMappingTest.kt:13]: Duplicate import 😄

Comment (chornerman) [reply template/data/src/test/java/co/nimblehq/template/data/extensions/ResponseMappingTest.kt:13]: @doannimble Did you update it yet? I don't seem to see any change 🤔

**Review by chornerman — APPROVED** [review]
LGTM 🚀

**Comment by Tuubz** [comment]
@doannimble Should we clean up the commit history with the correct issue references? It's quite messy 🙇 

Once cleaned up, we can merge this pull request 🚀

**Pull request merged**
